src/kbc_modules/cc_kbc/mod.rs

```diff
@@ -94,4 +120,60 @@ impl Kbc {
             .ok_or_else(|| anyhow!("TEE rsa key missing"))?;
         response.decrypt_output(key)
     }
+
+    fn tee(&self) -> String {
+        self.tee.clone()
+    }
+
+    fn kbs_uri(&self) -> String {
+        self.kbs_uri.clone()
+    }
+
+    fn kbs_client(&mut self) -> Result<&mut reqwest::Client> {
+        if let Some(kbs_client) = &mut self.kbs_client {
+            Ok(kbs_client)
+        } else {
+            Err(anyhow!("Start kbs http client failed"))
+        }
+    }
+
+    async fn request_kbs_resource(&mut self, resource_url: String) -> Result<Response> {
+        let mut res = self.kbs_client()?.get(&resource_url).send().await?;
+
+        if res.status() == reqwest::StatusCode::UNAUTHORIZED {
+            let kbs_uri = self.kbs_uri();
+
+            let auth_req = Request::new(self.tee());
+            let challenge = self.kbs_client()?
+                .post(format!("{}/auth", kbs_uri))
+                .header("Content-Type", "application/json")
+                .body(serde_json::to_string(&auth_req)?)
+                .send()
+                .await?
+                .json::<Challenge>()
+                .await?;
+            self.nonce = challenge.nonce.clone();
+
+            let attestation_evidence = self.generate_evidence()?;
+            let attest_response = self.kbs_client()?
+                .post(format!("{}/attest", kbs_uri))
+                .header("Content-Type", "application/json")
+                .body(serde_json::to_string(&attestation_evidence)?)
+                .send()
+                .await?;
+            if attest_response.status() != reqwest::StatusCode::OK {
+                // TODO: Parse Error description payload
+                return Err(anyhow!("KBS attest failed"));
+            }
```

Comment:
I suggest extracting this (/auth and /attest requests) to a separate fn (establish_session ?).

Reply:
Done.